client/dashboard/sites/site-leave-modal/index.tsx

```diff
@@ -0,0 +1,214 @@
+		confirmed: false,
+	} );
+
+	const isConfirmed = formData.confirmed;
+
+	const handleSubmit = ( e: React.FormEvent ) => {
+		e.preventDefault();
+		if ( ! me || ! isConfirmed ) {
+			return;
+		}
+
+		recordTracksEvent( 'calypso_sites_dashboard_site_leave_modal_leave_site_click' );
+
+		mutation.mutate( me.id, {
+			onSuccess: () => {
+				recordTracksEvent( 'calypso_sites_dashboard_site_leave_modal_leave_site_success' );
+				createSuccessNotice(
+					/* translators: %s: site domain */
+					sprintf( __( 'You have left %s successfully.' ), site.slug ),
+					{ type: 'snackbar' }
+				);
+
+				router.navigate( { to: '/sites' } );
+				onClose();
+			},
+			onError: ( error: Error ) => {
+				recordTracksEvent( 'calypso_sites_dashboard_site_leave_modal_leave_site_error', {
+					site_id: site.ID,
+					error: error.name,
+				} );
+
+				createErrorNotice( error.message, { type: 'snackbar' } );
+				onClose();
+			},
+		} );
+	};
+
+	return (
+		<form onSubmit={ handleSubmit }>
+			<VStack spacing={ 6 }>
+				<Text as="p">
+					{ createInterpolateElement(
+						/* translators: %s: site domain */
+						__( 'Are you sure to leave the site <siteDomain />?' ),
+						{
+							siteDomain: <strong>{ site.slug }</strong>,
+						}
+					) }
+				</Text>
+				<Text as="p">
+					{ __(
+						'Leaving will remove your access to the site, including all content, users, domains, upgrades, and anything else you have access to.'
+					) }
+				</Text>
+				<Text as="p">
+					{ __(
+						'To regain access, a current administrator must re-invite you. Please confirm this is your intent before proceeding.'
+					) }
+				</Text>
+				<DataForm< { confirmed: boolean } >
+					data={ formData }
+					fields={ [
+						{
+							id: 'confirmed',
+							label: __( 'I understand the consequences of leaving' ),
+							Edit: 'checkbox',
+						},
+					] }
+					form={ { type: 'regular' as const, fields: [ 'confirmed' ] } }
+					onChange={ ( edits: { confirmed?: boolean } ) => {
+						setFormData( ( data ) => ( { ...data, ...edits } ) );
+					} }
+				/>
+				<HStack spacing={ 4 } justify="flex-end" expanded={ false }>
+					<Button variant="tertiary" onClick={ onClose }>
+						{ __( 'Cancel' ) }
+					</Button>
+					<Button
+						variant="primary"
+						type="submit"
+						disabled={ ! isConfirmed }
+						isBusy={ mutation.isPending }
+					>
+						{ __( 'Leave site' ) }
+					</Button>
+				</HStack>
+			</VStack>
+		</form>
+	);
+}
+
+export default function SiteLeaveModal( { site, onClose }: SiteLeaveModalProps ) {
+	const { user } = useAuth();
+	const { data: hasPurchasesCancelable } = useQuery(
+		siteHasPurchasesCancelableQuery( site.slug, user.ID )
+	);
+
+	const showContentHasPurchasesCancelable = hasPurchasesCancelable;
+	const showContentSiteOwner = ! showContentHasPurchasesCancelable && isSiteOwner( user, site );
+	const showContentLeaveSite = ! showContentHasPurchasesCancelable && ! showContentSiteOwner;
```

Comment:
[nit] This is basically `if-else`, right? I think something like this is more readable...

```js
const renderContent = () => {
  if (hasPurchasesCancelable) {
    return <ContentHasPurchasesCancelable/>;
  }
  if (isSiteOwner(user, site)) {
    return <ContentSiteOwner/>;
  }
  return <ContentLeaveSite/>;
};
```

Reply:
Good idea, updated in 65e10c1d1dab310ab422d46ef8ac7d507f4edd90.